### PR TITLE
Enable nmodl as submodule to CoreNEURON

### PR DESCRIPTION
**Pull request by pramodk**
- can use following command to enable submodule:
    cmake .. -DCORENRN_ENABLE_NMODL=ON -DCMAKE_INSTALL_PREFIX=`pwd`/install -DCORENRN_ENABLE_ISPC=OFF
    cmake .. -DCORENRN_ENABLE_NMODL=ON -DCMAKE_INSTALL_PREFIX=`pwd`/install -DCORENRN_ENABLE_ISPC=ON
  - one can also set -DCORENRN_NMODL_DIR=<nmodl_install_prefix>
  - -DCORENRN_ENABLE_NMODL=OFF should use old mod2c
  - Change _ROOT to _DIR for consistency with NEURON CMake

Todo:
- [x] Once nmodl PR is merged, change submodule commit from cmake branch to master of nmodl : https://github.com/BlueBrain/nmodl/pull/245

### CODE REVIEW

**Comment by ohm314** [comment]
Just for clarification:
in
```
cmake .. -DCORENRN_ENABLE_NMODL=ON -DCMAKE_INSTALL_PREFIX=pwd/install -DCORENRN_ENABLE_ISPC=OFF
```
the `-DCMAKE_INSTALL_PREFIX` is not actually necessary, right?

File: CMakeLists.txt [hunk 1]
```diff
@@ -149,17 +149,24 @@ endif()
...
-  find_package(nmodl REQUIRED)
-  include_directories(${nmodl_INCLUDE})
+  find_package(nmodl)
 
-  # copy to make nrnivmodl-core work from build as well instal directory
-  configure_file(${nmodl_BINARY} ${CMAKE_BINARY_DIR}/bin/nmodl COPYONLY)
-  set(CORENRN_NMODL_BINARY ${CMAKE_BINARY_DIR}/bin/nmodl)
+  if (nmodl_FOUND)
+    # copy to make nrnivmodl-core work from build as well install directory
+    configure_file(${nmodl_BINARY} ${CMAKE_BINARY_DIR}/bin/nmodl COPYONLY)
```
Comment (ohm314) [inline CMakeLists.txt]: I didn't see this before.. is this really necessary?
We really shouldn't have to copy around binaries to make things work...

File: extra/nrnivmodl_core_makefile.in [hunk 2]
```diff
@@ -45,7 +45,12 @@ CXX_LINK_SHARED = $(CXX) $(CXXFLAGS) @CMAKE_SHARED_LIBRARY_CREATE_CXX_FLAGS@ @CM
 
 ISPC = @CMAKE_ISPC_COMPILER@
 ISPC_FLAGS = @CMAKE_ISPC_FLAGS@
-ISPC_COMPILE = $(ISPC) $(ISPC_FLAGS) -I$(incdir) -I@nmodl_INCLUDE@
+# Note that CORENRN_NMODL_INCLUDE will point to build directory (if nmodl is build as submodule)
```
Comment (ohm314) [inline extra/nrnivmodl_core_makefile.in]:
```suggestion
# Note that CORENRN_NMODL_INCLUDE will point to build directory (if nmodl is built as submodule)
```
</nitpick>

File: CMakeLists.txt [hunk 3]
```diff
@@ -63,8 +63,8 @@ option(CORENRN_ENABLE_CUDA_UNIFIED_MEMORY "Enable CUDA unified memory support" O
 option(CORENRN_ENABLE_UNIT_TESTS "Enable unit tests execution" ON)
 option(CORENRN_ENABLE_GPU "Enable GPU support using OpenACC" OFF)
 
-set(CORENRN_NMODL_ROOT "" CACHE PATH "Path nmodl source-to-source compiler root")
-set(LIKWID_ROOT "" CACHE PATH "Path to likwid performance analysis suite")
+set(CORENRN_NMODL_DIR "" CACHE PATH "Path to nmodl source-to-source compiler installation")
```
[the remark this answers is not in the document]
Comment (pramodk) [reply CMakeLists.txt]: done https://github.com/BlueBrain/spack/pull/579

File: CMakeLists.txt (hunk 1 above)
Comment (pramodk) [reply CMakeLists.txt]: agree! fixed this as per our discussion.

File: CMake/AddNmodlSubmodule.cmake [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+find_path(NMODL_PROJ NAMES CMakeLists.txt PATHS "${CORENEURON_PROJECT_SOURCE_DIR}/external/nmodl")
+find_package_handle_standard_args(NMODL REQUIRED_VARS NMODL_PROJ)
+
+if(NOT NMODL_FOUND)
+  find_package(Git 1.8.3 QUIET)
+  if(NOT ${GIT_FOUND})
+    message(FATAL_ERROR "git not found, clone repository with --recursive")
+  endif()
+  message(STATUS "Sub-module nmodl missing : running git submodule update --init --recursive")
+  execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive --
```
Comment (iomaganaris) [inline CMake/AddNmodlSubmodule.cmake]: This is super cool 👍

File: CMake/AddNmodlSubmodule.cmake [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+                          ${CORENEURON_PROJECT_SOURCE_DIR}/external/nmodl
+                  WORKING_DIRECTORY ${CORENEURON_PROJECT_SOURCE_DIR})
+else()
+  message(STATUS "Using nmodl submodule from ${NMODL_PROJ}")
+endif()
+
+add_subdirectory(${CORENEURON_PROJECT_SOURCE_DIR}/external/nmodl)
```
[the remark this answers is not in the document]
Comment (pramodk) [reply CMake/AddNmodlSubmodule.cmake]: done

**Comment by pramodk** [comment]
> They take some and I don't know if it's that useful to run with CoreNeuron

yeah, agree! I changed now that in nmodl.

**Pull request merged**